ctru-rs/src/services/ir_user.rs

```diff
@@ -0,0 +1,491 @@
+        expected_response_len: usize,
+    ) -> crate::Result<Vec<u32>> {
+        let mut shared_mem_guard = IR_USER_STATE.lock().unwrap();
+        let shared_mem = shared_mem_guard.as_mut().unwrap();
+
+        // Set up the request
+        let cmd_buffer = unsafe {
+            &mut *(slice_from_raw_parts_mut(
+                ctru_sys::getThreadCommandBuffer(),
+                max(request.len(), expected_response_len),
+            ))
+        };
+        cmd_buffer[0..request.len()].copy_from_slice(&request);
+
+        // Send the request
+        unsafe {
+            ResultCode(ctru_sys::svcSendSyncRequest(shared_mem.service_handle))?;
+        }
+
+        // Handle the result returned by the service
+        ResultCode(cmd_buffer[1] as ctru_sys::Result)?;
+
+        // Copy back the response
+        request.clear();
+        request.extend_from_slice(&cmd_buffer[0..expected_response_len]);
+
+        Ok(request)
+    }
+}
+
+// Internal helper for rounding up a value to a multiple of another value.
+fn round_up(value: usize, multiple: usize) -> usize {
+    if value % multiple != 0 {
+        (value / multiple) * multiple + multiple
+    } else {
+        (value / multiple) * multiple
+    }
+}
+
+struct InitializeIrnopSharedParams {
+    ir_user_handle: Handle,
+    shared_memory_len: u32,
+    recv_packet_buffer_len: u32,
+    recv_packet_count: u32,
+    send_packet_buffer_len: u32,
+    send_packet_count: u32,
+    bit_rate: u32,
+    shared_memory_handle: Handle,
+}
+
+/// Internal helper for initializing the ir:USER service
+unsafe fn initialize_irnop_shared(params: InitializeIrnopSharedParams) -> crate::Result<()> {
+    let cmd_buffer = &mut *(slice_from_raw_parts_mut(ctru_sys::getThreadCommandBuffer(), 9));
+    cmd_buffer[0] = INITIALIZE_IRNOP_SHARED_COMMAND_HEADER;
+    cmd_buffer[1] = params.shared_memory_len;
+    cmd_buffer[2] = params.recv_packet_buffer_len;
+    cmd_buffer[3] = params.recv_packet_count;
+    cmd_buffer[4] = params.send_packet_buffer_len;
+    cmd_buffer[5] = params.send_packet_count;
+    cmd_buffer[6] = params.bit_rate;
+    cmd_buffer[7] = 0;
+    cmd_buffer[8] = params.shared_memory_handle;
+
+    // Send the request
+    ResultCode(ctru_sys::svcSendSyncRequest(params.ir_user_handle))?;
+
+    // Handle the result returned by the service
+    ResultCode(cmd_buffer[1] as ctru_sys::Result)?;
+
+    Ok(())
+}
+
+/// An enum which represents the different IR devices the 3DS can connect to via
+/// the ir:USER service.
+pub enum IrDeviceId {
+    CirclePadPro,
+    // Pretty sure no other IDs are recognized, but just in case
+    Custom(u32),
+}
+
+impl IrDeviceId {
+    /// Get the ID of the device.
+    pub fn get_id(&self) -> u32 {
+        match *self {
+            IrDeviceId::CirclePadPro => 1,
+            IrDeviceId::Custom(id) => id,
+        }
+    }
+}
+
+/// This struct holds a parsed copy of the ir:USER service status (from shared memory).
+#[derive(Debug)]
+pub struct IrUserStatusInfo {
```

Comment:
Do these structs need to have `#[repr(C)]` to be compatible with the underlying OS and/or `libctru`?

Reply:
No, this is a copy of the data after parsing.